File: old/binary_search/first_occurrence_left_most_index_of_target/first_occurrence_left_most_index_of_target.py
import random


def binary_search_left_first_occurance(arr, val):
    left = 0
    right = len(arr) - 1
    lowest_index = -1

    while left <= right:
        middle = left + (right - left) // 2
        if arr[middle] < val:
            left = middle + 1
        elif arr[middle] > val:
            right = middle - 1
        else:
            lowest_index = middle
            right = middle - 1

    return lowest_index


def main():
    arr = [1, 2, 2, 3, 3, 3, 4, 5, 6, 7, 7, 8, 8, 8]
    val = random.randint(1, 10)
    index = binary_search_left_first_occurance(arr, val)
    print(arr)
    if index == -1:
        print("not found")
        return
    else:
        print(f"found {val} in index: {index}")


main()

File: old/binary_search/first_occurrence_left_most_index_of_target/first_occurrence_left_most_index_of_target.cpp
#include <iostream>
#include <vector>

using namespace std;

int binarySearchFirstOccurrence(vector<int>& arr, int val) {
    int left = 0;
    int right = arr.size() - 1;
    int currIndex = -1;

    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (arr[middle] < val) {
            left = middle + 1;
        } else if (arr[middle] > val) {
            right = middle - 1;
        } else {
            currIndex = middle;
            right = middle - 1;
        }

    }
    return currIndex;
}

int main() {
    vector<int> sortedArr = {1, 2, 2, 3, 3, 3, 4, 5, 6, 7, 8, 8, 8, 8, 9};
    for (int val: sortedArr) {
        cout << val << " ";
    }
    cout << endl;
    int findVal = 100;
    int index = binarySearchFirstOccurrence(sortedArr, findVal);
    if (index == -1) {
        cout << "not found" << endl;
        return 0;
    }

    cout << "first index that we found the val: " << findVal << " is in index: " << index << endl;
    return 0;
}
